### ISSUE TITLE
Generating controlled gate from QuantumCircuit results in AttributeError when circuit contains at least one explicit identity operator

### ISSUE THREAD

**Issue by JoelHBierman** [Qiskit/qiskit-terra#7399]
### Environment

- **Qiskit Terra version**: 0.19.1 (Also tested this in 0.18.0 with same outcome)
- **Python version**: 3.9.7
- **Operating system**: CentOS (Also tested this on MacOS 12.0.1 with same outcome) 

### What is happening?

When you take a `QuantumCircuit` where one of its constituent gates is explicitly the identity and you attempt to generate a controlled gate from this unitary using the `control()` method, Qiskit raises the error:

`AttributeError: 'NoneType' object has no attribute 'global_phase'`

### How can we reproduce the issue?

Examples of code that reproduce this error:

```
from qiskit.opflow.operator_globals import X, Y, Z, I

pauli_str_op = Z^Z^Z^I

pauli_str_op.to_circuit().control()
```

```
from qiskit import Aer, QuantumCircuit, QuantumRegister, AncillaRegister, ClassicalRegister
from qiskit.utils.algorithm_globals import algorithm_globals
from qiskit.utils.backend_utils import is_aer_provider
from qiskit.opflow.operator_globals import X, Y, Z, I


pauli_str_op = Z^Z^Z^I

qreg = QuantumRegister(4, 'q')
circ = QuantumCircuit(qreg)
circ.z(qreg[0])
circ.z(qreg[1])
circ.z(qreg[2])
circ.id(qreg[3])

circ.control()
```

This results in the following Traceback:

```
Traceback (most recent call last):
  File "/Users/joelbierman/Desktop/Qiskit_Projects/control_gate_error.py", line 6, in <module>
    pauli_str_op.to_circuit().control()
  File "/Users/joelbierman/opt/anaconda3/envs/Qiskit_0280_pyscf2_ARM64/lib/python3.9/site-packages/qiskit/circuit/quantumcircuit.py", line 554, in control
    controlled_gate = gate.control(num_ctrl_qubits, label, ctrl_state)
  File "/Users/joelbierman/opt/anaconda3/envs/Qiskit_0280_pyscf2_ARM64/lib/python3.9/site-packages/qiskit/circuit/gate.py", line 119, in control
    return add_control(self, num_ctrl_qubits, label, ctrl_state)
  File "/Users/joelbierman/opt/anaconda3/envs/Qiskit_0280_pyscf2_ARM64/lib/python3.9/site-packages/qiskit/circuit/add_control.py", line 59, in add_control
    cgate = control(operation, num_ctrl_qubits=num_ctrl_qubits, label=label, ctrl_state=ctrl_state)
  File "/Users/joelbierman/opt/anaconda3/envs/Qiskit_0280_pyscf2_ARM64/lib/python3.9/site-packages/qiskit/circuit/add_control.py", line 113, in control
    unrolled_gate = _unroll_gate(operation, basis_gates=basis)
  File "/Users/joelbierman/opt/anaconda3/envs/Qiskit_0280_pyscf2_ARM64/lib/python3.9/site-packages/qiskit/circuit/add_control.py", line 268, in _unroll_gate
    opqc = dag_to_circuit(unroller.run(dag))
  File "/Users/joelbierman/opt/anaconda3/envs/Qiskit_0280_pyscf2_ARM64/lib/python3.9/site-packages/qiskit/transpiler/passes/basis/unroller.py", line 113, in run
    unrolled_dag = self.run(decomposition)  # recursively unroll ops
  File "/Users/joelbierman/opt/anaconda3/envs/Qiskit_0280_pyscf2_ARM64/lib/python3.9/site-packages/qiskit/transpiler/passes/basis/unroller.py", line 76, in run
    raise QiskitError(
qiskit.exceptions.QiskitError: "Error decomposing node of instruction 'id': 'NoneType' object has no attribute 'global_phase'. Unable to define instruction 'id' in the given basis."
```

### What should happen?

Controlled unitary of `QuantumCircuit` is generated, regardless of what unitary gates make up the circuit.

### Any suggestions?

This seems to be related to how Qiskit handles generating controlled unitaries from circuits that explicitly contain at least one identity operator. This of course does not happen if the identity implicitly acts on one of the qubits since this would, by definition, break the `control()` method for all circuits. For example, the following code does not raise this error:

```
from qiskit import Aer, QuantumCircuit, QuantumRegister, AncillaRegister, ClassicalRegister
from qiskit.utils.algorithm_globals import algorithm_globals
from qiskit.utils.backend_utils import is_aer_provider
from qiskit.opflow.operator_globals import X, Y, Z, I


pauli_str_op = Z^Z^Z^I

qreg = QuantumRegister(4, 'q')
circ = QuantumCircuit(qreg)
circ.z(qreg[0])
circ.z(qreg[1])
circ.z(qreg[2])
# nothing acts on qubit 3: the identity is implicit.
circ.control()

```

**Comment by jakelishman** [comment]
This is because `IGate`, which we use to describe the explicit identity operation, doesn't have a `_define` method.  The fix is _probably_ just to define in `qiskit/circuit/library/standard_gates/i.py`:
```python
class IGate(Gate):
    ...
    
    def _define(self):
        self.definition = QuantumCircuit(1, name=self.name)
```
but I wouldn't like to swear that won't have a couple of knock-on issues in the transpiler, or affect performance.

**Comment by JoelHBierman** [comment]
I see. Thank you!

**Comment by JoelHBierman** [comment]
I can confirm that this seems to remove the AttributeError. Of course, I haven't done any testing to see if it breaks something else or if it results in reduced performance.

**Comment by Cryoris** [comment]
We didn't see any reason why opflow should actually insert an `IGate` for Pauli identity operations, especially since they are eagerly removed upon calling `reduce` on any operator expression. Actually, I think inserting an `IGate` is even wrong since it currently represent a waiting cycle in the circuit, but we expect a mathematical identity operation: nothing should happen!

#7569 changes this behavior and as by-product fixes the first example shown in this issue:
```python
from qiskit.opflow.operator_globals import X, Y, Z, I

pauli_str_op = Z^Z^Z^I

pauli_str_op.to_circuit().control()  # works with 7569 since not IGate is contained
```
Now I'm not sure if we can close this issue if the linked PR is merged, since it only covers one of the described cases.

**Comment by JoelHBierman** [comment]
I will just note that my original motivation for posting this issue was to fix the Pauli string op issue specifically. It could be that others will also care about the second case getting fixed. I think if the linked PR is merged, this issue can be closed since the second case is mentioned in #4010.

**Comment by jakelishman** [comment]
It's probably best to leave the issue open, since the other part of it is still going to be an issue after this, until we're able to realign `IGate` to be more what most people expect.  But it's good that Julien's patch will fix your actual issue.

**Comment by jakelishman** [comment]
We're going to defer the issue til the next minor release - we'll merge the immediate fix (#7569) into the `main` branch soon, but there's a risk that will break other code currently in use elsewhere in Qiskit or for users, so we'll wait til 0.20 before it gets released.

**Comment by kdk** [comment]
Since #7569 has merged, I'm going to remove this issue from the 0.20 milestone. Is there anything here that's not covered by #4010 , #3950 and #7403 ? If so, can someone clarify what remains and update the title?

**Comment by jakelishman** [comment]
I think the remaining problems raised by this issue are just duplicates of the ones that Kevin's raised, so I'll close this now.

**Comment by lironmo** [comment]
it's seems that i am getting similar error, i am using the following version  (latest):
qiskit==0.37.1
qiskit-aer==0.10.4
qiskit-ibmq-provider==0.19.2
qiskit-terra==0.21.1

i simple create 4 random circuit and add control for all of them:

`
random.seed(102030)
np.random.seed(102030)

def create_circuit_list(num_circuits, circuit_num_qubits, max_depth):
    """Generate random array of circuits of arbitrary size and form.

      Args:
          num_circuits (int): number of quantum circuits to create
          circuit_num_qubits (int): number of quantum wires
          max_depth (int): max layers of operations (i.e. critical path length)

      Returns:
          List: constructed circuits list

      Raises:
          IError: when invalid options given
      """
    get_logger(LOGGER_NAME).debug(
        f"Enter num_circuits=[{num_circuits}], circuit_num_qubits=[{circuit_num_qubits}], max_depth={max_depth}")
    assert max_depth >= 0
    ret_circuits = [random_circuit(num_qubits=circuit_num_qubits, depth=random.randint(1, max_depth)) for i in
                    range(num_circuits)]
    return ret_circuits


def add_qiskit_control_list(circuits_list):
    ret_circuits = [circuit.control() for circuit in circuits_list]
    return ret_circuits
`

repreduce step:
- fix seed (see above)
- run the following:
`
    num_circuits = 3
    circuit_num_qubits = 4
    max_depth = 3
    circuits = create_circuit_list(num_circuits, circuit_num_qubits, max_depth)
    circuits = add_qiskit_control_list(circuits)
`

error log:

2022-08-02 18:03:02,740 [SELECT:DEBUG], [24408:MainThread] [select_u_implementation.py:<module>:53] | Start DEBUGGING
2022-08-02 18:03:02,740 [SELECT:DEBUG], [24408:MainThread] [select_u_implementation.py:create_circuit_list:29] | Enter num_circuits=[3], circuit_num_qubits=[4], max_depth=3
Traceback (most recent call last):
  File "C:\envs\qiskit\lib\site-packages\qiskit\transpiler\passes\basis\unroller.py", line 82, in run
    phase = node.op.definition.global_phase
AttributeError: 'NoneType' object has no attribute 'global_phase'

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "C:\phd\code\qiskit\select_u_implementation.py", line 58, in <module>
    circuits = add_qiskit_control_list(circuits)
  File "C:\phd\code\qiskit\select_u_implementation.py", line 38, in add_qiskit_control_list
    ret_circuits = [circuit.control() for circuit in circuits_list]
  File "C:\phd\code\qiskit\select_u_implementation.py", line 38, in <listcomp>
    ret_circuits = [circuit.control() for circuit in circuits_list]
  File "C:\envs\qiskit\lib\site-packages\qiskit\circuit\quantumcircuit.py", line 706, in control
    controlled_gate = gate.control(num_ctrl_qubits, label, ctrl_state)
  File "C:\envs\qiskit\lib\site-packages\qiskit\circuit\gate.py", line 118, in control
    return add_control(self, num_ctrl_qubits, label, ctrl_state)
  File "C:\envs\qiskit\lib\site-packages\qiskit\circuit\add_control.py", line 59, in add_control
    cgate = control(operation, num_ctrl_qubits=num_ctrl_qubits, label=label, ctrl_state=ctrl_state)
  File "C:\envs\qiskit\lib\site-packages\qiskit\circuit\add_control.py", line 113, in control
    unrolled_gate = _unroll_gate(operation, basis_gates=basis)
  File "C:\envs\qiskit\lib\site-packages\qiskit\circuit\add_control.py", line 269, in _unroll_gate
    opqc = dag_to_circuit(unroller.run(dag))
  File "C:\envs\qiskit\lib\site-packages\qiskit\transpiler\passes\basis\unroller.py", line 85, in run
    raise QiskitError(
qiskit.exceptions.QiskitError: "Error decomposing node of instruction 'id': 'NoneType' object has no attribute 'global_phase'. Unable to define instruction 'id' in the given basis."

Process finished with exit code 1

**Comment by lironmo** [comment]
@jakelishman this issue seems to be closed but i am using the latest package and get the same error.

do you have a workaround?

**Comment by jakelishman** [comment]
We don't have a fix yet, sorry, this issue is just closed because it's a duplicate of #4010.  It's not at all ideal, but in your use case, you could either modify your `random_circuit` function so it never produces `IGate` instances, or you could write a wrapper function that looks something like this:
```python
from qiskit import QuantumCircuit
from qiskit.circuit.library import IGate

def my_random_circuit(qubits, depth):
    base = random_circuit(qubits, depth)
    qc = base.copy_empty_like()
    for instruction in base:
        if isinstance(instruction.operation, IGate):
            continue
        qc.append(instruction)
    return qc
```

That just uses the output of `random_circuit`, and throws away any `id` gates in it.

**Comment by lironmo** [comment]
Thanks for the detailed response!
(Pretty much what I did.)

Side question, it seems that the qiskit implementation of the control is very inefficient with depth increasing exponentially in the number of controls
